### PR TITLE
implemented exchange toggler

### PR DESCRIPTION
**Pull request by Sirmorrison**
fixes issue #87
Issues with the selectors are resolved in PR 102 as this pr only addresses issues with exchange toggling.

image

image

### CODE REVIEW

**Comment by Sirmorrison** [comment]
> 
> 
> image
> Is shown whenever bluetrade is selected

image
resolved

File: postgres/exchange_ticks.go [hunk 1]
```diff
@@ -258,14 +257,19 @@ func (pg *PgDb) AllExchangeTicksInterval(ctx context.Context) ([]ticks.TickDtoIn
 	return TickDtoInterval, err
 }
 
-func (pg *PgDb) ChartExchangeTicks(ctx context.Context, selectedDtick string, currencyPair string, selectedInterval int) ([]ticks.TickChart, error) {
-	exchangeFilterResult, err := models.ExchangeTicks(qm.Select(fmt.Sprintf("%s, time, exchange_id", selectedDtick)), qm.Where("currency_pair=? and interval=?", currencyPair, selectedInterval), qm.OrderBy(models.ExchangeTickColumns.Time)).All(ctx, pg.db)
+func (pg *PgDb) ChartExchangeTicks(ctx context.Context, selectedDtick string, currencyPair string, selectedInterval int, source string) ([]ticks.TickChart, error) {
```
Comment (ademuanthony) [inline postgres/exchange_ticks.go]: what is `selectedDtick`? `selectedTick`?

File: postgres/exchange_ticks.go [hunk 2]
```diff
@@ -258,14 +257,19 @@ func (pg *PgDb) AllExchangeTicksInterval(ctx context.Context) ([]ticks.TickDtoIn
...
+	exchange, err := models.Exchanges(models.ExchangeWhere.Name.EQ(source)).One(ctx, pg.db)
+	if err != nil {
+		return nil, err
+	}
+
+	exchangeFilterResult, err := models.ExchangeTicks(qm.Select(fmt.Sprintf("%s, time, exchange_id", selectedDtick)), qm.Where("currency_pair=? and interval=? and exchange_id=?", currencyPair, selectedInterval, exchange.ID), qm.OrderBy(models.ExchangeTickColumns.Time)).All(ctx, pg.db)
```
Comment (ademuanthony) [inline postgres/exchange_ticks.go]: let's use the fields from the column name struct and remove the hard-coded fields; `time` and `exchange_id`

File: postgres/exchange_ticks.go [hunk 3]
```diff
@@ -258,14 +257,19 @@ func (pg *PgDb) AllExchangeTicksInterval(ctx context.Context) ([]ticks.TickDtoIn
...
 	if err != nil {
 		return nil, err
 	}
 
-	tickChart := []ticks.TickChart{}
 	var Filter float64
+	tickChart := []ticks.TickChart{}
```
Comment (ademuanthony) [inline postgres/exchange_ticks.go]: `tickChartData`, I think. Rename the struct `TickChart` to `TickChartData` as well

File: web/handlers.go [hunk 4]
```diff
@@ -43,12 +43,28 @@ func (s *Server) getExchangeTicks(res http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	// intervalFilter := make(map[string]int, 0)
```
Comment (ademuanthony) [inline web/handlers.go]: remove the unused line of code

File: web/handlers.go [hunk 5]
```diff
@@ -43,12 +43,28 @@ func (s *Server) getExchangeTicks(res http.ResponseWriter, req *http.Request) {
...
 	intervals, err := s.db.AllExchangeTicksInterval(ctx)
 	if err != nil {
 		s.renderError(err.Error(), res)
 		return
 	}
 
+	// for _, val := range intervals {
```
Comment (ademuanthony) [inline web/handlers.go]: remove the unused lines of code

File: web/handlers.go [hunk 6]
```diff
@@ -156,16 +175,19 @@ func (s *Server) getChartData(res http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	chartData, err := s.db.ChartExchangeTicks(ctx, selectedDtick, selectedCpair, interval)
+	chartData, err := s.db.ChartExchangeTicks(ctx, selectedDtick, selectedCpair, interval, sources)
```
Comment (ademuanthony) [inline web/handlers.go]: rename `s.db.ChartExchangeTicks` to `s.db.ExchangeTicksChartData`

File: web/handlers.go [hunk 7]
```diff
@@ -156,16 +175,19 @@ func (s *Server) getChartData(res http.ResponseWriter, req *http.Request) {
...
 	if err != nil {
 		s.renderError(err.Error(), res)
 		return
 	}
-
-	data := map[string]interface{}{
-		"chartData": chartData,
+	if len(chartData) == 0{
+		data["message"] = fmt.Sprintf("Not enough Data to generate %s chart.", sources)
```
Comment (ademuanthony) [inline web/handlers.go]: is it that there is `Not enough data` or there is `no data` at all?

Comment (ademuanthony) [inline web/handlers.go]: it is not that there is `Not enough data`, the issue is there is `no data` at all?

File: web/public/app/src/controllers/exchange_controller.js [hunk 8]
```diff
@@ -11,7 +11,8 @@ export default class extends Controller {
       'selectedFilter', 'exchangeTable', 'selectedCpair', 'numPageWrapper', 'intervalWapper',
```
Comment (ademuanthony) [inline web/public/app/src/controllers/exchange_controller.js]: `selectedCpair` should be `selectedCurrencyPair`

File: web/public/app/src/controllers/exchange_controller.js [hunk 9]
```diff
@@ -11,7 +11,8 @@ export default class extends Controller {
...
       'previousPageButton', 'totalPageCount', 'nextPageButton', 'selectedDticks', 'selectedInterval',
       'exRowTemplate', 'currentPage', 'selectedNum', 'exchangeTableWrapper', 'tickWapper',
-      'chartWrapper', 'labels', 'chartsView', 'viewOption', 'hideOption', 'sourceWrapper'
+      'chartWrapper', 'labels', 'chartsView', 'viewOption', 'hideOption', 'sourceWrapper',
+      'pageSizeWrapper', 'chartSource', 'cPHideOption', 'messageView'
```
Comment (ademuanthony) [inline web/public/app/src/controllers/exchange_controller.js]: `cPHideOption`, do you mean `currencyPairHideOption`?

File: web/public/app/src/controllers/exchange_controller.js [hunk 10]
```diff
@@ -24,35 +25,36 @@ export default class extends Controller {
...
+    hide(this.cPHideOptionTarget)
+    hide(this.numPageWrapperTarget)
+    hide(this.exchangeTableWrapperTarget)
     this.setActiveOptionBtn(opt, this.viewOptionTargets)
-    this.chartWrapperTarget.classList.remove('d-hide')
-    var y = this.selectedIntervalTarget.options
-    this.selectedInterval = this.selectedIntervalTarget.value = y[0].text
+    this.selectedInterval = this.selectedIntervalTarget.value = interval[0].value
+    this.selectedFilter = this.selectedFilterTarget.value = sFilter[1].text
     this.selectedDtick = this.selectedDticksTarget.value = 'close'
```
Comment (ademuanthony) [inline web/public/app/src/controllers/exchange_controller.js]: please rename `this.selectedDtick` and `this.selectedDtick`

File: web/public/app/src/controllers/exchange_controller.js [hunk 11]
```diff
@@ -171,49 +167,68 @@ export default class extends Controller {
...
-      item,
-      key
-
-    for (i = 0; i < dataSet.length; i++) {
-      item = dataSet[i]
-      key = item[0].toString()
-      if (!hash[key]) {
-        hash[key] = item.slice()
-        continue
+      // var hash = {}
```
Comment (ademuanthony) [inline web/public/app/src/controllers/exchange_controller.js]: remove unused lines of code

File: web/public/app/src/controllers/exchange_controller.js [hunk 12]
```diff
@@ -171,49 +167,68 @@ export default class extends Controller {
...
+      //   if (!hash[key]) {
+      //     hash[key] = item.slice()
+      //     continue
+      //   }
+      //   for (j = 1; j < item.length; j++) hash[key][j] += item[j]
+      // }
+
+      // result = Object.values(hash)
+
+      let labels = ['Date']
```
Comment (ademuanthony) [inline web/public/app/src/controllers/exchange_controller.js]: `let labels = ['Date']` can be written as `let labels = ['Date', _this.selectedFilter]`
`let colors = ['#007bff']` can be written as `let colors = ['#007bff', getRandomColor()]`
then line 206 and 207 can be removed

File: web/public/app/src/controllers/exchange_controller.js [hunk 13]
```diff
@@ -171,49 +167,68 @@ export default class extends Controller {
...
-    }
 
-    result = Object.values(hash)
-
-    console.log(result)
-    _this.chartsView = new Dygraph(
-      _this.chartsViewTarget,
-      result, { ...options, ...extra }
-    )
+      console.log(dataSet)
```
Comment (ademuanthony) [inline web/public/app/src/controllers/exchange_controller.js]: remove debug helper

File: web/server.go [hunk 14]
```diff
@@ -25,7 +25,7 @@ type DataQuery interface {
 	AllExchange(ctx context.Context) (models.ExchangeSlice, error)
 	FetchExchangeTicks(ctx context.Context, currencyPair, name string, offset int, limit int) ([]ticks.TickDto, int64, error)
 	AllExchangeTicksCurrencyPair(ctx context.Context) ([]ticks.TickDtoCP, error)
-	ChartExchangeTicks(ctx context.Context, filter string, currencyPair string, selectedInterval int) ([]ticks.TickChart, error)
+	ChartExchangeTicks(ctx context.Context, filter string, currencyPair string, selectedInterval int, exchanges string) ([]ticks.TickChart, error)
```
Comment (ademuanthony) [inline web/server.go]: `ExchangeTicksChartData`

File: web/views/exchange.html [hunk 15]
```diff
@@ -34,38 +34,39 @@
...
                             </select>
                         </div>
-                        <div class="col-md-4 control-div">
+                        <div class="col-md-4 control-div p-0">
                             <div class="control-label">Currency Pair:</div>
                             <select data-target="exchange.selectedCpair" data-action="change->exchange#selectedCpairChanged"
-                                class="form-control" style="width: 110px;">
-                                <option data-target="exchange.hideOption value="All">All</option>
+                                class="form-control" style="width: 105px;">
+                                <option data-target="exchange.cPHideOption" value="All">All</option>
```
Comment (ademuanthony) [inline web/views/exchange.html]: `cPHideOption`, note the refactoring requested earlier

File: web/handlers.go (hunk 7 above)
Comment (Sirmorrison) [reply web/handlers.go]: not enough data sounds more appropriate to me.

**Pull request merged**